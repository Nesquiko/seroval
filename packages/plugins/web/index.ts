export { default as BlobPlugin } from './blob';
export { default as FilePlugin } from './file';
export { default as HeadersPlugin } from './headers';
export { default as URLSearchParamsPlugin } from './url-search-params';
export { default as URL } from './url';
