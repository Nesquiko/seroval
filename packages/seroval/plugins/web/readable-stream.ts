import type { SerovalNode, Stream } from 'seroval';
import { createPlugin, createStream } from 'seroval';

function toStream<T>(value: ReadableStream<T>): Stream<T | undefined> {
  const stream = createStream<T | undefined>();

  const reader = value.getReader();

  async function push(): Promise<void> {
    try {
      const result = await reader.read();
      if (result.done) {
        stream.return(result.value);
      } else {
        stream.next(result.value);
        await push();
      }
    } catch (error) {
      stream.throw(error);
    }
  }

  push().catch(() => {
    //
  });

  return stream;
}

interface ReadableStreamNode {
  stream: SerovalNode;
}

export default createPlugin<ReadableStream, ReadableStreamNode>({
  tag: 'seroval/plugins/web/readable-stream',
  test(value) {
    if (typeof ReadableStream === 'undefined') {
      return false;
    }
    return value instanceof ReadableStream;
  },
  parse: {
    sync(value, ctx, data) {
      return {
        stream: ctx.parse(createStream()),
      };
    },
    async async(value, ctx, data) {
      return {
        stream: await ctx.parse(toStream(value)),
      };
    },
    stream(value, ctx, data) {
      return {
        stream: ctx.parse(toStream(value)),
      };
    },
  },
  serialize(node, ctx, data) {

  },
  deserialize(node, ctx, data) {

  },
});
