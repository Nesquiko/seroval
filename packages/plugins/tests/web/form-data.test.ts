import { describe, it, expect } from 'vitest';
import {
  crossSerialize,
  crossSerializeAsync,
  crossSerializeStream,
  deserialize,
  fromCrossJSON,
  fromJSON,
  serialize,
  serializeAsync,
  toCrossJSON,
  toCrossJSONAsync,
  toCrossJSONStream,
  toJSON,
  toJSONAsync,
} from 'seroval';
import { FormDataPlugin } from '../../web';

const SYNC_EXAMPLE = new FormData();
SYNC_EXAMPLE.set('hello', 'world');
SYNC_EXAMPLE.set('foo', 'bar');

const ASYNC_EXAMPLE = new FormData();
ASYNC_EXAMPLE.set('hello-world', new File(['Hello World'], 'hello.txt', {
  type: 'text/plain',
  lastModified: 1681027542680,
}));
ASYNC_EXAMPLE.set('foo-bar', new File(['Foo Bar'], 'foo-bar.txt', {
  type: 'text/plain',
  lastModified: 1681027542680,
}));

describe('FormData', () => {
  describe('serialize', () => {
    it('supports FormData', () => {
      const result = serialize(SYNC_EXAMPLE, {
        plugins: [FormDataPlugin],
      });
      expect(result).toMatchSnapshot();
      const back = deserialize<typeof SYNC_EXAMPLE>(result);
      expect(back).toBeInstanceOf(FormData);
    });
  });
  describe('serializeAsync', () => {
    it('supports FormData', async () => {
      const result = await serializeAsync(ASYNC_EXAMPLE, {
        plugins: [FormDataPlugin],
      });
      expect(result).toMatchSnapshot();
      const back = deserialize<typeof ASYNC_EXAMPLE>(result);
      expect(back).toBeInstanceOf(FormData);
    });
  });
  describe('toJSON', () => {
    it('supports FormData', () => {
      const result = toJSON(SYNC_EXAMPLE, {
        plugins: [FormDataPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<typeof SYNC_EXAMPLE>(result, {
        plugins: [FormDataPlugin],
      });
      expect(back).toBeInstanceOf(FormData);
    });
  });
  describe('toJSONAsync', () => {
    it('supports FormData', async () => {
      const result = await toJSONAsync(ASYNC_EXAMPLE, {
        plugins: [FormDataPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<typeof ASYNC_EXAMPLE>(result, {
        plugins: [FormDataPlugin],
      });
      expect(back).toBeInstanceOf(FormData);
    });
  });
  describe('crossSerialize', () => {
    it('supports FormData', () => {
      const result = crossSerialize(SYNC_EXAMPLE, {
        plugins: [FormDataPlugin],
      });
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports FormData', () => {
        const result = crossSerialize(SYNC_EXAMPLE, {
          plugins: [FormDataPlugin],
          scopeId: 'example',
        });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeAsync', () => {
    it('supports FormData', async () => {
      const result = await crossSerializeAsync(ASYNC_EXAMPLE, {
        plugins: [FormDataPlugin],
      });
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports FormData', async () => {
        const result = await crossSerializeAsync(ASYNC_EXAMPLE, {
          plugins: [FormDataPlugin],
          scopeId: 'example',
        });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeStream', () => {
    it('supports FormData', async () => new Promise<void>((resolve, reject) => {
      crossSerializeStream(Promise.resolve(SYNC_EXAMPLE), {
        plugins: [FormDataPlugin],
        onSerialize(data) {
          expect(data).toMatchSnapshot();
        },
        onDone() {
          resolve();
        },
        onError(error) {
          reject(error);
        },
      });
    }));
    describe('scoped', () => {
      it('supports FormData', async () => new Promise<void>((resolve, reject) => {
        crossSerializeStream(Promise.resolve(SYNC_EXAMPLE), {
          plugins: [FormDataPlugin],
          scopeId: 'example',
          onSerialize(data) {
            expect(data).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
    });
  });
  describe('toCrossJSON', () => {
    it('supports FormData', () => {
      const result = toCrossJSON(SYNC_EXAMPLE, {
        plugins: [FormDataPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromCrossJSON<typeof SYNC_EXAMPLE>(result, {
        plugins: [FormDataPlugin],
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(FormData);
    });
  });
  describe('toCrossJSONAsync', () => {
    it('supports FormData', async () => {
      const result = await toCrossJSONAsync(ASYNC_EXAMPLE, {
        plugins: [FormDataPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromCrossJSON<typeof ASYNC_EXAMPLE>(result, {
        plugins: [FormDataPlugin],
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(FormData);
    });
  });
  describe('toCrossJSONStream', () => {
    it('supports FormData', async () => new Promise<void>((resolve, reject) => {
      toCrossJSONStream(Promise.resolve(SYNC_EXAMPLE), {
        plugins: [FormDataPlugin],
        onParse(data) {
          expect(JSON.stringify(data)).toMatchSnapshot();
        },
        onDone() {
          resolve();
        },
        onError(error) {
          reject(error);
        },
      });
    }));
  });
});
