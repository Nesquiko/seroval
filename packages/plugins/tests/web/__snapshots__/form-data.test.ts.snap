// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormData > crossSerialize > scoped > supports FormData 1`] = `"($R=>$R[0]=($R[1]=(e,f,i,s,t)=>{f=new FormData;for(i=0,s=e.length;i<s;i++)f.append((t=e[i])[0],t[1]);return f})($R[2]=[$R[3]=["hello","world"],$R[4]=["foo","bar"]]))($R["example"])"`;

exports[`FormData > crossSerialize > supports FormData 1`] = `"$R[0]=($R[1]=(e,f,i,s,t)=>{f=new FormData;for(i=0,s=e.length;i<s;i++)f.append((t=e[i])[0],t[1]);return f})($R[2]=[$R[3]=["hello","world"],$R[4]=["foo","bar"]])"`;

exports[`FormData > crossSerializeAsync > scoped > supports FormData 1`] = `"($R=>$R[0]=($R[1]=(e,f,i,s,t)=>{f=new FormData;for(i=0,s=e.length;i<s;i++)f.append((t=e[i])[0],t[1]);return f})($R[2]=[$R[3]=["hello-world",$R[4]=new File([$R[6]=new Uint8Array([72,101,108,108,111,32,87,111,114,108,100]).buffer],"hello.txt",$R[5]={type:"text/plain",lastModified:1681027542680})],$R[7]=["foo-bar",$R[8]=new File([$R[10]=new Uint8Array([70,111,111,32,66,97,114]).buffer],"foo-bar.txt",$R[9]={type:"text/plain",lastModified:1681027542680})]]))($R["example"])"`;

exports[`FormData > crossSerializeAsync > supports FormData 1`] = `"$R[0]=($R[1]=(e,f,i,s,t)=>{f=new FormData;for(i=0,s=e.length;i<s;i++)f.append((t=e[i])[0],t[1]);return f})($R[2]=[$R[3]=["hello-world",$R[4]=new File([$R[6]=new Uint8Array([72,101,108,108,111,32,87,111,114,108,100]).buffer],"hello.txt",$R[5]={type:"text/plain",lastModified:1681027542680})],$R[7]=["foo-bar",$R[8]=new File([$R[10]=new Uint8Array([70,111,111,32,66,97,114]).buffer],"foo-bar.txt",$R[9]={type:"text/plain",lastModified:1681027542680})]])"`;

exports[`FormData > crossSerializeStream > scoped > supports FormData 1`] = `"($R=>$R[0]=($R[1]=(s,f,p)=>((p=new Promise((a,b)=>{s=a,f=b})).s=s,p.f=f,p))())($R["example"])"`;

exports[`FormData > crossSerializeStream > scoped > supports FormData 2`] = `"($R=>($R[7]=(p,d)=>{p.s(d),p.status="success",p.value=d;delete p.s;delete p.f})($R[0],$R[2]=($R[3]=(e,f,i,s,t)=>{f=new FormData;for(i=0,s=e.length;i<s;i++)f.append((t=e[i])[0],t[1]);return f})($R[4]=[$R[5]=["hello","world"],$R[6]=["foo","bar"]])))($R["example"])"`;

exports[`FormData > crossSerializeStream > supports FormData 1`] = `"$R[0]=($R[1]=(s,f,p)=>((p=new Promise((a,b)=>{s=a,f=b})).s=s,p.f=f,p))()"`;

exports[`FormData > crossSerializeStream > supports FormData 2`] = `"($R[7]=(p,d)=>{p.s(d),p.status="success",p.value=d;delete p.s;delete p.f})($R[0],$R[2]=($R[3]=(e,f,i,s,t)=>{f=new FormData;for(i=0,s=e.length;i<s;i++)f.append((t=e[i])[0],t[1]);return f})($R[4]=[$R[5]=["hello","world"],$R[6]=["foo","bar"]]))"`;

exports[`FormData > serialize > supports FormData 1`] = `"((e,f,i,s,t)=>{f=new FormData;for(i=0,s=e.length;i<s;i++)f.append((t=e[i])[0],t[1]);return f})([["hello","world"],["foo","bar"]])"`;

exports[`FormData > serializeAsync > supports FormData 1`] = `"((e,f,i,s,t)=>{f=new FormData;for(i=0,s=e.length;i<s;i++)f.append((t=e[i])[0],t[1]);return f})([["hello-world",new File([new Uint8Array([72,101,108,108,111,32,87,111,114,108,100]).buffer],"hello.txt",{type:"text/plain",lastModified:1681027542680})],["foo-bar",new File([new Uint8Array([70,111,111,32,66,97,114]).buffer],"foo-bar.txt",{type:"text/plain",lastModified:1681027542680})]])"`;

exports[`FormData > toCrossJSON > supports FormData 1`] = `"{"t":25,"i":0,"s":{"factory":{"t":25,"i":1,"c":"seroval-plugins/web/FormDataFactory"},"entries":{"t":9,"i":2,"l":2,"a":[{"t":9,"i":3,"l":2,"a":[{"t":1,"s":"hello"},{"t":1,"s":"world"}],"o":0},{"t":9,"i":4,"l":2,"a":[{"t":1,"s":"foo"},{"t":1,"s":"bar"}],"o":0}],"o":0}},"c":"seroval-plugins/web/FormData"}"`;

exports[`FormData > toCrossJSONAsync > supports FormData 1`] = `"{"t":25,"i":0,"s":{"factory":{"t":25,"i":1,"c":"seroval-plugins/web/FormDataFactory"},"entries":{"t":9,"i":2,"l":2,"a":[{"t":9,"i":3,"l":2,"a":[{"t":1,"s":"hello-world"},{"t":25,"i":4,"s":{"name":{"t":1,"s":"hello.txt"},"options":{"t":10,"i":5,"p":{"k":["type","lastModified"],"v":[{"t":1,"s":"text/plain"},{"t":0,"s":1681027542680}],"s":2},"o":0},"buffer":{"t":19,"i":6,"s":[72,101,108,108,111,32,87,111,114,108,100]}},"c":"seroval-plugins/web/File"}],"o":0},{"t":9,"i":7,"l":2,"a":[{"t":1,"s":"foo-bar"},{"t":25,"i":8,"s":{"name":{"t":1,"s":"foo-bar.txt"},"options":{"t":10,"i":9,"p":{"k":["type","lastModified"],"v":[{"t":1,"s":"text/plain"},{"t":0,"s":1681027542680}],"s":2},"o":0},"buffer":{"t":19,"i":10,"s":[70,111,111,32,66,97,114]}},"c":"seroval-plugins/web/File"}],"o":0}],"o":0}},"c":"seroval-plugins/web/FormData"}"`;

exports[`FormData > toCrossJSONStream > supports FormData 1`] = `"{"t":22,"i":0,"f":{"t":26,"i":1,"s":1}}"`;

exports[`FormData > toCrossJSONStream > supports FormData 2`] = `"{"t":23,"i":0,"a":[{"t":26,"i":7,"s":2},{"t":25,"i":2,"s":{"factory":{"t":25,"i":3,"c":"seroval-plugins/web/FormDataFactory"},"entries":{"t":9,"i":4,"l":2,"a":[{"t":9,"i":5,"l":2,"a":[{"t":1,"s":"hello"},{"t":1,"s":"world"}],"o":0},{"t":9,"i":6,"l":2,"a":[{"t":1,"s":"foo"},{"t":1,"s":"bar"}],"o":0}],"o":0}},"c":"seroval-plugins/web/FormData"}]}"`;

exports[`FormData > toJSON > supports FormData 1`] = `"{"t":{"t":25,"i":0,"s":{"factory":{"t":25,"i":1,"c":"seroval-plugins/web/FormDataFactory"},"entries":{"t":9,"i":2,"l":2,"a":[{"t":9,"i":3,"l":2,"a":[{"t":1,"s":"hello"},{"t":1,"s":"world"}],"o":0},{"t":9,"i":4,"l":2,"a":[{"t":1,"s":"foo"},{"t":1,"s":"bar"}],"o":0}],"o":0}},"c":"seroval-plugins/web/FormData"},"f":31,"m":[]}"`;

exports[`FormData > toJSONAsync > supports FormData 1`] = `"{"t":{"t":25,"i":0,"s":{"factory":{"t":25,"i":1,"c":"seroval-plugins/web/FormDataFactory"},"entries":{"t":9,"i":2,"l":2,"a":[{"t":9,"i":3,"l":2,"a":[{"t":1,"s":"hello-world"},{"t":25,"i":4,"s":{"name":{"t":1,"s":"hello.txt"},"options":{"t":10,"i":5,"p":{"k":["type","lastModified"],"v":[{"t":1,"s":"text/plain"},{"t":0,"s":1681027542680}],"s":2},"o":0},"buffer":{"t":19,"i":6,"s":[72,101,108,108,111,32,87,111,114,108,100]}},"c":"seroval-plugins/web/File"}],"o":0},{"t":9,"i":7,"l":2,"a":[{"t":1,"s":"foo-bar"},{"t":25,"i":8,"s":{"name":{"t":1,"s":"foo-bar.txt"},"options":{"t":10,"i":9,"p":{"k":["type","lastModified"],"v":[{"t":1,"s":"text/plain"},{"t":0,"s":1681027542680}],"s":2},"o":0},"buffer":{"t":19,"i":10,"s":[70,111,111,32,66,97,114]}},"c":"seroval-plugins/web/File"}],"o":0}],"o":0}},"c":"seroval-plugins/web/FormData"},"f":31,"m":[]}"`;
